### PR TITLE
feat: inject app service for charm retrieval into legacy state

### PR DESCRIPTION
**Pull request by nvinuesa**
This patch wires the domain services for charm retrieval into the legacy state. It is an important milestone because it allows us to start cleaning up the legacy state regarding charms.

Many small changes were needed to adapt the output of the legacy charm retrieval with the new domain and all its call sites, among them:
- `Version()` was added to the interfaces **and** as a new field in the CharmInfo returned RPC params.
- `IsAvailable` was added as a new return value on `GetCharm()`. This was needed for the object store to return the correct error when the charm was already available.

Also, `UpdateUploadedCharm()` was removed since no longer needed now that we are reading from dqlite.

## QA steps

Bootstrap and deploy ubuntu should continue working. We are aware of local charms not working, and more complex scenarios will be tested and fixed when we continue removing state.

```
$ juju bootstrap lxd c
$ juju add-model m
$ juju deploy ubuntu
```

## Links

**Jira card:** JUJU-6846

### CODE REVIEW

File: state/application.go [hunk 1]
```diff
@@ -3493,19 +3482,7 @@ func (a *Application) UnitNames() ([]string, error) {
...
-	if origin == nil {
-		return false
-	}
-	// The charm may be downloaded, but the application's
-	// data may not updated yet. This can happen when multiple
-	// applications share a charm.
-	notReady := origin.Source == "charm-hub" && origin.ID == ""
-	return !ch.IsPlaceholder() && !ch.IsUploaded() || notReady
+	return true
 }
```
Comment (jack-w-shaw) [inline state/application.go:3486]: This is about to disappear

File: internal/charm/services/storage.go
Comment (jack-w-shaw) [inline internal/charm/services/storage.go:1]: This whole package is about to disappear

**Review by jack-w-shaw — APPROVED** [review]

**Review by SimonRichardson — APPROVED** [review]

**Pull request merged**